--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1633,7 +1633,7 @@ static int airoha_dev_open(struct net_de
 
 #if defined(CONFIG_PCS_AIROHA)
 	if (airhoa_is_phy_external(port)) {
-		err = phylink_of_phy_connect(port->phylink,
+		err = phylink_of_phy_connect(dev->phylink,
 					     netdev->dev.of_node, 0);
 		if (err) {
 			netdev_err(netdev, "%s: could not attach PHY: %d\n", __func__,
@@ -1641,7 +1641,7 @@ static int airoha_dev_open(struct net_de
 			return err;
 		}
 
-		phylink_start(port->phylink);
+		phylink_start(dev->phylink);
 	}
 #endif
 
@@ -1712,8 +1712,8 @@ static int airoha_dev_stop(struct net_de
 
 #if defined(CONFIG_PCS_AIROHA)
 	if (airhoa_is_phy_external(port)) {
-		phylink_stop(port->phylink);
-		phylink_disconnect_phy(port->phylink);
+		phylink_stop(dev->phylink);
+		phylink_disconnect_phy(dev->phylink);
 	}
 #endif
 
@@ -2867,10 +2867,10 @@ static const struct ethtool_ops airoha_e
 static struct phylink_pcs *airoha_phylink_mac_select_pcs(struct phylink_config *config,
 							 phy_interface_t interface)
 {
-	struct airoha_gdm_port *port = container_of(config, struct airoha_gdm_port,
-						    phylink_config);
+	struct airoha_gdm_dev *dev = container_of(config, struct airoha_gdm_dev,
+						  phylink_config);
 
-	return port->pcs;
+	return dev->pcs;
 }
 
 static void airoha_mac_config(struct phylink_config *config, unsigned int mode,
@@ -2936,8 +2936,9 @@ static void airoha_mac_link_up(struct ph
 			       unsigned int mode, phy_interface_t interface,
 			       int speed, int duplex, bool tx_pause, bool rx_pause)
 {
-	struct airoha_gdm_port *port = container_of(config, struct airoha_gdm_port,
-						    phylink_config);
+	struct airoha_gdm_dev *dev = container_of(config, struct airoha_gdm_dev,
+						  phylink_config);
+	struct airoha_gdm_port *port = dev->port;
 	struct airoha_qdma *qdma = port->qdma;
 	struct airoha_eth *eth = qdma->eth;
 	u32 frag_size_tx, frag_size_rx;
@@ -2983,7 +2984,6 @@ static int airoha_setup_phylink(struct n
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct device_node *np = netdev->dev.of_node;
-	struct airoha_gdm_port *port = dev->port;
 	phy_interface_t phy_mode;
 	struct phylink *phylink;
 	int err;
@@ -2994,34 +2994,34 @@ static int airoha_setup_phylink(struct n
 		return err;
 	}
 
-	port->phylink_config.dev = &netdev->dev;
-	port->phylink_config.type = PHYLINK_NETDEV;
-	port->phylink_config.mac_capabilities = MAC_ASYM_PAUSE | MAC_SYM_PAUSE |
-						MAC_10 | MAC_100 | MAC_1000 | MAC_2500FD |
-						MAC_5000FD | MAC_10000FD;
+	dev->phylink_config.dev = &netdev->dev;
+	dev->phylink_config.type = PHYLINK_NETDEV;
+	dev->phylink_config.mac_capabilities = MAC_ASYM_PAUSE | MAC_SYM_PAUSE |
+					       MAC_10 | MAC_100 | MAC_1000 |
+					       MAC_2500FD | MAC_5000FD |
+					       MAC_10000FD;
 
 	__set_bit(PHY_INTERFACE_MODE_SGMII,
-		  port->phylink_config.supported_interfaces);
+		  dev->phylink_config.supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_1000BASEX,
-		  port->phylink_config.supported_interfaces);
+		  dev->phylink_config.supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_2500BASEX,
-		  port->phylink_config.supported_interfaces);
+		  dev->phylink_config.supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_USXGMII,
-		  port->phylink_config.supported_interfaces);
+		  dev->phylink_config.supported_interfaces);
 	__set_bit(PHY_INTERFACE_MODE_10GBASER,
-		  port->phylink_config.supported_interfaces);
+		  dev->phylink_config.supported_interfaces);
 
-	port->pcs = airoha_pcs_create(&netdev->dev);
-	if (IS_ERR(port->pcs))
-		return PTR_ERR(port->pcs);
+	dev->pcs = airoha_pcs_create(&netdev->dev);
+	if (IS_ERR(dev->pcs))
+		return PTR_ERR(dev->pcs);
 
-	phylink = phylink_create(&port->phylink_config,
-				 of_fwnode_handle(np),
+	phylink = phylink_create(&dev->phylink_config, of_fwnode_handle(np),
 				 phy_mode, &airoha_phylink_ops);
 	if (IS_ERR(phylink))
 		return PTR_ERR(phylink);
 
-	port->phylink = phylink;
+	dev->phylink = phylink;
 
 	return 0;
 }
@@ -3307,18 +3307,17 @@ error_hw_cleanup:
 			continue;
 
 		airoha_metadata_dst_free(port);
-#if defined(CONFIG_PCS_AIROHA)
-		if (port->phylink) {
-			phylink_destroy(port->phylink);
-			airoha_pcs_destroy(port->pcs);
-		}
-#endif
 		for (j = 0; j < ARRAY_SIZE(port->devs); j++) {
 			struct airoha_gdm_dev *dev = port->devs[j];
 
 			if (!dev)
 				continue;
-
+#if defined(CONFIG_PCS_AIROHA)
+			if (dev->phylink) {
+				phylink_destroy(dev->phylink);
+				airoha_pcs_destroy(dev->pcs);
+			}
+#endif
 			if (dev->dev->reg_state == NETREG_REGISTERED)
 				unregister_netdev(dev->dev);
 		}
@@ -3347,19 +3346,17 @@ static void airoha_remove(struct platfor
 			continue;
 
 		airoha_metadata_dst_free(port);
-#if defined(CONFIG_PCS_AIROHA)
-		if (port->phylink) {
-			phylink_destroy(port->phylink);
-			airoha_pcs_destroy(port->pcs);
-		}
-#endif
-
 		for (j = 0; j < ARRAY_SIZE(port->devs); j++) {
 			struct airoha_gdm_dev *dev = port->devs[j];
 
 			if (!dev)
 				continue;
-
+#if defined(CONFIG_PCS_AIROHA)
+		if (dev->phylink) {
+			phylink_destroy(dev->phylink);
+			airoha_pcs_destroy(dev->pcs);
+		}
+#endif
 			airoha_dev_stop(dev->dev);
 			unregister_netdev(dev->dev);
 		}
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -536,6 +536,12 @@ struct airoha_gdm_dev {
 	struct airoha_gdm_port *port;
 	struct net_device *dev;
 
+#if defined(CONFIG_PCS_AIROHA)
+	struct phylink *phylink;
+	struct phylink_config phylink_config;
+	struct phylink_pcs *pcs;
+#endif
+
 	int nbq;
 };
 
@@ -544,12 +550,6 @@ struct airoha_gdm_port {
 	struct airoha_gdm_dev *devs[AIROHA_MAX_NUM_GDM_DEVS];
 	int id;
 
-#if defined(CONFIG_PCS_AIROHA)
-	struct phylink *phylink;
-	struct phylink_config phylink_config;
-	struct phylink_pcs *pcs;
-#endif
-
 	struct airoha_hw_stats stats;
 
 	DECLARE_BITMAP(qos_sq_bmap, AIROHA_NUM_QOS_CHANNELS);
